File: ruby/anagram/anagram.rb
class Anagram
  def initialize(word)
    @word = Anagram.normalized_case(word)
    @chars = Anagram.sorted_chars(@word)
  end

  def self.normalized_case(str)
    str.downcase
  end

  def self.sorted_chars(str)
    str.chars.sort
  end

  def match(candidates)
    candidates.select { |candidate| anagram?(candidate) }
  end

  private

  attr_reader :word, :chars

  def anagram?(candidate)
    candidate = Anagram.normalized_case(candidate)

    return false if candidate == word # words are not anagrams of themselves

    Anagram.sorted_chars(candidate) == chars
  end
end

File: go/anagram/anagram.go
package anagram

import (
	"sort"
	"strings"
)

func Detect(subject string, candidates []string) []string {
	subject = strings.ToLower(subject)

	var anagrams []string
	for _, candidate := range candidates {
		if isAnagram(subject, strings.ToLower(candidate)) {
			anagrams = append(anagrams, candidate)
		}
	}
	return anagrams
}

func isAnagram(subject, candidate string) bool {
	if len(subject) != len(candidate) {
		return false
	}

	if subject == candidate {
		return false
	}

	return sortedString(subject) == sortedString(candidate)
}

func sortedString(s string) string {
	bytes := []byte(s)
	sort.Slice(bytes, func(i, j int) bool {
		return bytes[i] < bytes[j]
	})
	return string(bytes)
}
